### PR TITLE
added top_k and top_p sampling techniques

### PR DESCRIPTION
**Pull request by paudelnirajan**
* top k tokens to choose from
* top p (nucleus sampling)

### CODE REVIEW

**Review by look-luc — COMMENTED** [review]
👍

File: src/spec_decode.py [hunk 1]
```diff
@@ -376,3 +383,38 @@ def speculative_decode_different_tokenizers():
...
+    indices_to_remove = logits < kth_value
+    logits_filtered = logits.masked_fill(indices_to_remove, float('-inf'))
+
+    return logits_filtered
+
+def apply_top_p(logits: torch.Tensor, p: float) -> torch.Tensor:
+    """Filters logits to keep the smallest set of top tokens whose cumulative prob >= p."""
+
+    if p >= 1.0:
+        return logits
```
Comment (michaelpginn) [inline src/spec_decode.py:420]: Might be better to raise an error so someone doesn't accidentally use like 95.0 and it fails silently

**Review by michaelpginn — APPROVED** [review]
LGTM, maybe just change to throwing an error

File: src/spec_decode.py (hunk 1 above)
Comment (paudelnirajan) [reply src/spec_decode.py:420]: added checks for valid `p` and `k`.

File: src/spec_decode.py [hunk 2]
```diff
@@ -127,7 +131,7 @@ def speculative_decode(
         device = next(target_model.parameters()).device
 
     def select_index(logits: torch.Tensor):
-        return sample(logits, mode)
+        return sample(logits, mode, top_k=top_k, top_p=top_p)
```
[the remark this answers is not in the document]
Comment (michaelpginn) [reply src/spec_decode.py]: @paudelnirajan This is right---can we move the logic filtering logic out of the sampling function, and instead apply a filter as soon as we get the logits?

**Review by michaelpginn — CHANGES_REQUESTED** [review]

File: src/spec_decode.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (paudelnirajan) [reply src/spec_decode.py]: Done.

**Review by michaelpginn — APPROVED** [review]
